### ISSUE TITLE
Incompatible Kotlin version

### ISSUE THREAD

**Issue by bogdanzurac** [ioki-mobility/LokaliseGradlePlugin#98]
Hi. We're getting the following error with the latest `2.3.0-SNAPSHOT` update from yesterday:

```
Class 'com.ioki.lokalise.gradle.plugin.LokaliseApiFactory' was compiled with an incompatible version of Kotlin. The actual metadata version is 2.1.0, but the compiler version 1.9.0 can read versions up to 2.0.0.
The class is loaded from /Users/x/.gradle/caches/modules-2/files-2.1/com.ioki.lokalise/lokalise-gradle-plugin/2.3.0-SNAPSHOT/9c7107311ce3f346f0672183f01cfc0d948da0ba/lokalise-gradle-plugin-2.3.0-SNAPSHOT.jar!/com/ioki/lokalise/gradle/plugin/LokaliseApiFactory.class
```

This used to work with the previous versions of `2.3.0-SNAPSHOT` between 9th November and 12th December. But it seems you have uploaded a new `2.3.0-SNAPSHOT` version yesterday, on 17th December, which caused this issue: https://s01.oss.sonatype.org/content/repositories/snapshots/com/ioki/lokalise/com.ioki.lokalise.gradle.plugin/2.3.0-SNAPSHOT/

Our project uses Kotlin `2.0.10`, which is why we can't use Lokalise Gradle Plugin `2.2.0`. We've also tried with Kotlin `2.1.0`, but the issue is still present.

**Comment by StefMa** [comment]
Hey,
Thanks for the report.
Yesterday I merged Kotlin 2.1. See https://github.com/ioki-mobility/LokaliseGradlePlugin/commit/e025d3677a489d49a473731bffa056aa09539d32
Each merge to main creates a new snapshot version, thst is why you're facing that problem now.

When you say "You also tried with 2.1, same problem", what do you mean by that? Do you get the same issue? Or another/similar one?
If it's the same, then maybe it is a caching problem?

**Comment by StefMa** [comment]
I just checked your error again and I am a bit confused. You say you're using kotlin 2.0.20. But the error message reveals that you're using 1.9.

Furthermore, kotlin has a backward compatibility guarantee. So if u use kotlin 2.0, the current stable plugin should be usable for you, as it is compiled with Kotlin 1.9.

**Comment by StefMa** [comment]
Last one 😅 
I created a branch here https://github.com/ioki-mobility/LokaliseGradlePlugin/tree/kotlin-version-conflict that contains a `consumer`. That might look like your configuration.
However, I can't reproduce the error.
It compiles just fine and works as expected.
If I do 
```
cd consumer && ./gradlew downloadTranslationsForStrings
```
I get the expected error
```
* What went wrong:
Execution failed for task ':downloadTranslationsForStrings'.
> Can't download files
  Invalid `X-Api-Token` header
```

Let me know if you setup is different.

**Comment by bogdanzurac** [comment]
Our setup is a bit more complex, as we're creating our own custom plugin by using code from within your plugin (in this case, specifically `com.ioki.lokalise.gradle.plugin.LokaliseApiFactory`).

**Comment by StefMa** [comment]
Alright, you're using our project as a dependency 🙃 

I updated the [branch](https://github.com/ioki-mobility/LokaliseGradlePlugin/compare/main...kotlin-version-conflict) (https://github.com/ioki-mobility/LokaliseGradlePlugin/commit/41b2453d47c14bf07ce3d13be37df29a79856759)

When using Kotlin 1.9.0 with the current 2.3.0-SNAPSHOT I indeed get the same error.
However, when I upgrade to 2.0.10 "it build fine".
So that version should work for you 🤔 

but whats really strange is, that you can't use the current stable plugin (2.2.0) with any Kotlin version.
The error reports are also super strange..
```
 Class 'com.ioki.lokalise.gradle.plugin.LokaliseApiFactory' is compiled by a pre-release version of Kotlin and cannot be loaded by this version of the compiler
```

**Comment by StefMa** [comment]
I might found a workaround for the current stable plugin (2.2.0).
If you add
```kotlin
tasks{
    withType<org.jetbrains.kotlin.gradle.tasks.KotlinCompile> {
        compilerOptions {
            freeCompilerArgs.set(listOf("-Xskip-prerelease-check", "-Xallow-unstable-dependencies"))
        }
    }
}
```
to your build script, it seems to work 👀 

Let me know if that helps!

**Comment by bogdanzurac** [comment]
Yep, I can confirm that the workaround works. However, `2.3.0-SNAPSHOT` doesn't seem to work with Kotlin `2.0.10` in our case.
Do you know what might cause the underlying issue itself?

**Comment by StefMa** [comment]
@bogdanzurac can we close this issue or is there still a request/report from you? 🤔

**Comment by bogdanzurac** [comment]
> I might found a workaround for the current stable plugin (2.2.0). If you add
> 
> tasks{
>     withType<org.jetbrains.kotlin.gradle.tasks.KotlinCompile> {
>         compilerOptions {
>             freeCompilerArgs.set(listOf("-Xskip-prerelease-check", "-Xallow-unstable-dependencies"))
>         }
>     }
> }
> to your build script, it seems to work 👀
> 
> Let me know if that helps!

I can confirm it works with this workaround for `2.2.0`, but I still think this should be fixed in the upcoming `2.3.0` release.

**Comment by StefMa** [comment]
Can you use the current stable version instead? 🤔
If not, I will rollback Kotlin to 2.0.20 or so, and do a new release.

Let me know if the first one would work for you.

**Comment by bogdanzurac** [comment]
Hello again @StefMa. We have done a full update of all dependencies in your app and now have Kotlin v2.2.21. This works with LokaliseGradlePlugin v2.2.0 + the workaround mentioned above and with `"-Xskip-prerelease-check", "-Xallow-unstable-dependencies"`.

However, we have also tried updating the LokaliseGradlePlugin to the recently published v2.3.0. Now it stopped working with the `"-Xskip-prerelease-check", "-Xallow-unstable-dependencies"` hack. It throws this error:

```
e: Incompatible classes were found in dependencies. Remove them from the classpath or use '-Xskip-metadata-version-check' to suppress errors
e: file:///Users/s-devops-macos/.gradle/caches/modules-2/files-2.1/com.ioki.lokalise/lokalise-gradle-plugin/2.3.0/2dba19aa7fcf0313084c4c560b5553de5eb0e8b5/lokalise-gradle-plugin-2.3.0.jar!/META-INF/LokaliseGradlePlugin.kotlin_module Module was compiled with an incompatible version of Kotlin. The binary version of its metadata is 2.2.0, expected version is 2.0.0.
...
e: file:///Users/s-devops-macos/agents/agent-2/_work/1/s/buildPlugins/src/main/kotlin/tasks/util/download.kt:4:40 Class 'com.ioki.lokalise.gradle.plugin.LokaliseApiFactory' was compiled with an incompatible version of Kotlin. The actual metadata version is 2.2.0, but the compiler version 2.0.0 can read versions up to 2.1.0.
The class is loaded from /Users/s-devops-macos/.gradle/caches/modules-2/files-2.1/com.ioki.lokalise/lokalise-gradle-plugin/2.3.0/2dba19aa7fcf0313084c4c560b5553de5eb0e8b5/lokalise-gradle-plugin-2.3.0.jar!/com/ioki/lokalise/gradle/plugin/LokaliseApiFactory.class
```

If however I switch the hack to use only `"-Xskip-metadata-version-check"`, then v2.3.0 works as well.

Do you think you can take a quick look again at this issue?

**Comment by StefMa** [comment]
Hy @bogdanzurac ,

sorry for the late reply.
Let us check this:
[This GradlePlugin@2.3.0 uses Kotlin 2.2.21](https://github.com/ioki-mobility/LokaliseGradlePlugin/blob/2.3.0/gradle/libs.versions.toml#L2).

You wrote 
> We have done a full update of **all dependencies** in your **app** and now have Kotlin v2.2.21

Be note that **app** (assuming an Android app or "the project itself") might have a different Kotlin version than the plugin uses! 

I'm not 100% sure how all of this works.
But maybe Gradle is also involed here.
They have here a nice overview: https://docs.gradle.org/9.2.1/userguide/compatibility.html#kotlin
Do you use Gradle 8.11 or 8.12 ? Because they ship with Kotlin 2.0.X.
Because your error messages reveals that:
>  Module was compiled with an incompatible version of Kotlin. The binary version of its metadata is 2.2.0, expected version is 2.0.0.
> [...]
>  The actual metadata version is 2.2.0, but the compiler version 2.0.0 can read versions up to 2.1.0.

For me it seems that you're using Kotlin 2.0.X. Which can read up to 2.1.X but not 2.2+.

Is an Gradle update possible for you? Maybe also just for testing purpose? 

Update:
See also https://cookbook.gradle.org/plugin-development/kotlin-plugins/#gradle-compatibility

**Comment by bogdanzurac** [comment]
Hi Stefan!
No worries for the late reply, thanks for following up!

I've checked the compatibility matrix between Gradle version and Kotlin version that you've shared and I've double checked, we have in fact not updated Gradle yet to 9.x. We were still using 8.14.3 due to some other dependencies blocking the update. We have updated it now to 9.2.1 and I can confirm that now both at Gradle sync time, as well as at runtime, everything works perfectly, without anymore compiler arg hacks! I think we can now close this issue down completely.

Sorry for the confusion and thanks again for all your time and help. Have a lovely day!

**Comment by StefMa** [comment]
No worries.
Thanks for the update 👍 
I'm here to help.
Let me know if you face any other issues – or the same again in the future 😉
